flocker/dockerplugin/_api.py

```diff
@@ -235,6 +237,24 @@ def got_path(path):
             getting_path.addCallback(got_path)
             return getting_path
         d.addCallback(get_state)
+
+        def timeout():
+            d.cancel()
+
+        delayed_timeout = self._reactor.callLater(self._MOUNT_TIMEOUT, timeout)
+
+        def abort_timeout(passthrough):
+            if delayed_timeout.active():
+                delayed_timeout.cancel()
+            return passthrough
+        d.addBoth(abort_timeout)
+
+        def handleCancel(failure):
+            failure.trap(CancelledError)
+            return {u"Err": u"Timed out waiting for dataset to mount.",
+                    u"Mountpoint": u""}
+
+        d.addErrback(handleCancel)
```

Comment:
So I'm slightly partial to using cancellation for the timeout because it works precisely even if the timeout is not a multiple of poll_interval, but I'm thinking that loop_until might make the tricky parts of testing better.